### PR TITLE
Add support for Adblock Plus domain lists

### PR DESCRIPTION
**Pull request by DL6ER**
# What does this implement/fix?

This PR implements support for AdBlock Plus (ABP)-style domain lists in Pi-hole. To be precise, it adds support for the following domain matching syntax defined [here](https://adblockplus.org/filter-cheatsheet#blocking2) at the time of opening this PR:

image

We do *not* implement any other features such as *exception rules* as they are typically beyond what a DNS server can do (path information, for instance, is simply not available).

It should be noted that this new feature is not for free but the rather complex syntax means it comes at some computational costs (= delays in DNS replies if you are on low-end hardware). To mitigate this drawback, ABP-style matching is only enabled when FTL actually detects such domains in the `gravity` table. This shouldn't be the case for the vast majority of users using "normal" HOSTS-style or simple one-domain-per-line adlists as sources for Pi-hole.

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
4. I have commented my proposed changes within the code.
6. I am willing to help maintain this change if there are issues with it later.
7. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
8. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

### CODE REVIEW

File: src/database/gravity-db.c [hunk 1]
```diff
@@ -1290,7 +1328,102 @@ enum db_result in_gravity(const char *domain, clientsData *client)
...
+		// ... and use it to create a "gap" of the right size in our ABP
+		// format buffer
+		// Insert the domain component into the gap
+		if(ptr[0] == '.')
+		{
+			// If the component starts with a dot, we need
+			// to skip it when copying it into the ABP buffer
+			// Move excluding initial "||" but including final \0 (strlen-2+1 = strlen-1)
+			memmove(abpDomain+2+component_size-1, abpDomain+2, strlen(abpDomain)-1);
+			// Copy component bytes (excl. trailtrailingling null-byte)
```
Comment (dschaper) [inline src/database/gravity-db.c:1384]:
```suggestion
			// Copy component bytes (excl. trailing null-byte)
```

**Comment by dschaper** [comment]
Will need a matching PR in core to allow gravity.sh to pass the new format.

**Comment by PromoFaux** [comment]
🎉

```
root@docker-pihole-pi4:/# pihole-FTL sqlite3 /etc/pihole/gravity.db "Select * from gravity limit 1"
||0.code.cotsta.ru^|2
root@docker-pihole-pi4:/# nslookup 0.code.cotsta.ru 127.0.0.1
Server:         127.0.0.1
Address:        127.0.0.1#53

Name:   0.code.cotsta.ru
Address: 0.0.0.0
Name:   0.code.cotsta.ru
Address: ::

root@docker-pihole-pi4:/# nslookup subdomain.0.code.cotsta.ru 127.0.0.1
Server:         127.0.0.1
Address:        127.0.0.1#53

Name:   subdomain.0.code.cotsta.ru
Address: 0.0.0.0
Name:   subdomain.0.code.cotsta.ru
Address: ::
```

**Comment by jpgpi250** [comment]
> It should be noted that this new feature is not for free but the rather complex syntax means it comes at some computational costs (= delays in DNS replies if you are on low-end hardware). To mitigate this drawback, ABP-style matching is only enabled when FTL actually detects such domains in the gravity table.

could you please ensure there is a warning when "ABP-style matching" is enabled, either in the FTL log or in the web interface ("pi-hole diagnosis"), this to ensure the system warns there might be a "computational cost".

Thanks.

**Comment by yubiuser** [comment]
I added an info to the gravity script

```
  [i] Target: https://small.oisd.nl
  [✓] Status: Retrieval successful
  [i] List contained AdBlock Plus style domains
  [i] Imported 72149 domains
  [i] List has been updated
```

**Comment by jpgpi250** [comment]
> I added an info to the gravity script

gravityupdate runs on sunday early morning, not too many people look at the resulting log. An entry in the FTL log **would be highly appreciated**, monitoring this log allows me to generate the appropriate alert.

**Comment by yubiuser** [comment]
Just for my understanding: what do you intend to do, if you would get such a warning about an adlist? What are the consequences of the warning?

**Comment by jpgpi250** [comment]
Given the predicted performance cost, I would remove the list that activates "ABP-style matching", I'm happy with the amount of blocking I achieve now (35.4%) and confident that the additional domains are already (or will be) part of the regular lists I currently use.
Hopefully, this the list, containing these entries is easy to find...

**Comment by DL6ER** [comment]
> Given the predicted performance cost, I would remove the list that activates "ABP-style matching

Having a certain high-performance computing background, my choice of words may have not been adequate. Typically, gravity lookups take on the order of few nanoseconds so with ABP style blocking enabled we're still talking sub-milliseconds here. I don't think it's worth making a big deal out of it but it also won't harm to add a log message like
```
gravity contains ABP-style domains
```

**Comment by jpgpi250** [comment]
> gravity contains ABP-style domains

+ list url and / or ID please

**Comment by jfb-pihole** [comment]
> * list url and / or ID please

Is this really necessary?  There may be multiple lists of this type, depending on what the user added.

**Comment by DL6ER** [comment]
No, I will neither add the list URL nor the adlist ID. The reason is that this would be a *very* expensive database query. Image *expensive* now on the order of several seconds to a few minutes of FTL downtime. The reason is that we'd have to do a full database scan to get this information. Currently, FTL scans only if there is *any* domain in ABP style and stops right when it found one. However, there may be multiple adlists in the database so we'd have to scan the entire table until we have collected all adlist IDs.

As always, you can easily get what you want by a third-party script running an appropriate SQL query, such as
``` sql
SELECT DISTINCT adlist_id FROM gravity WHERE domain LIKE '||%^';
```

**Comment by rdwebdesign** [comment]
The information is already on  `gravity.sh` here:

https://github.com/pi-hole/pi-hole/blob/821c7dc190cdabc28f024e4688bff3670f028731/gravity.sh#L539-L544

**Comment by DL6ER** [comment]
I pushed a new commit and set this PR back to draft because I'm interested in your findings about timing. Please, check out again this branch and have a look into your `FTL.log` file for something like
```
TIMED gravity_check_ABP_format(): .... ms
```

What does it say in your case? Do you have ABP domains in your gravity or not? I'm also interested in results once with and once without these

**Comment by yubiuser** [comment]
I tried this on a x86 (Celeron J4105 ) once with a single ABP list (https://small.oisd.nl)

```
TIMED gravity_check_ABP_format(): 325.687554 ms
```

and once without

```
TIMED gravity_check_ABP_format(): 323.809821 ms
```

**Comment by jpgpi250** [comment]
raspberry pi 3B
Linux raspberrypi 5.15.84-v7+ #1613 SMP Thu Jan 5 11:59:48 GMT 2023 armv7l GNU/Linux

**no ADB lists!**

pihole checkout ftl new/adb_style_blocking

pihole -v
  Pi-hole version is v5.15.5 (Latest: v5.15.5)
  AdminLTE version is v5.18.4 (Latest: v5.18.4)
  FTL version is new/adb_style_blocking vDev-ae9b291 (Latest: v5.21)

```
[2023-02-26 12:26:45.305 22265/T22281] TIMED gravity_check_ABP_format(): 0.126719 ms
```

I assume there is no "abp_domains" property in the "info" table of the gravity.db, because this requires also installing pihole core (database version change?).

**Comment by yubiuser** [comment]
> I assume there is no "abp_domains" property in the "info" table of the gravity.db, because this requires also installing pihole core (database version change?).

Yes.

https://github.com/pi-hole/pi-hole/pull/5179

**Comment by jpgpi250** [comment]
replaced gravity.sh and query.sh from github source

new list created for testing, daily updated, list contains known DoH server domains, ADB style.

```
  [i] Target: https://raw.githubusercontent.com/jpgpi250/piholemanual/master/DOH/DOHadb.txt
  [✓] Status: No changes detected
  [i] List contained AdBlock Plus style domains
  [i] Imported 2479 domains
```

pihole -q looks OK, example:
```
 pihole -q 1qaz.de
 Match found in http://localhost.localdomain/DOHservers/DOHadb.txt:
   ||1qaz.de^
```

gravity.db / info table / abp_domains -> **value = 1** (but the list contains 2479 domain entries), is this intended?

**Comment by DL6ER** [comment]
> gravity.db / info table / abp_domains -> **value = 1** (but the list contains 2479 domain entries), is this intended?

Yes, this is a `FALSE`/`TRUE` value stored by SQLite3 internally a [zero-byte integer](https://www.sqlite.org/fileformat2.html#record_format) to save space. SQLite3 maps `FALSE -> 0` and `TRUE -> 1`.

I removed the timing output now. A fraction of a millisecond fells way better than several hundred milliseconds, possibly more. I will not clean the history of this PR as we might want to have a look at some details later.

This PR is ready for review and merge.

**Comment by jpgpi250** [comment]
thanks, updated, will keep testing.

Word of warning: beware of the default separator '|' sqlite3 uses. Some of my scripts needed a rewrite to handle the ABP-style, e.g. gravity entries starting with '||'

**Comment by PromoFaux** [comment]
Yes

https://github.com/pi-hole/pi-hole/pull/5179/commits/c5faf3d1744f100bbbfc800702c70ffc79eda02e

File: src/database/gravity-db.c [hunk 2]
```diff
@@ -32,6 +32,9 @@
 // Definition of struct regexData
 #include "../regex_r.h"
 
+// CHECK_TIME()
```
Comment (yubiuser) [inline src/database/gravity-db.c:35]:
```suggestion

```

File: src/database/gravity-db.c [hunk 3]
```diff
@@ -32,6 +32,9 @@
...
+#include "../timers.h"
```
Comment (yubiuser) [inline src/database/gravity-db.c:36]:
```suggestion

```

File: src/database/gravity-db.c [hunk 4]
```diff
@@ -32,6 +32,9 @@
...
+
```
Comment (yubiuser) [inline src/database/gravity-db.c:37]:
```suggestion
```

**Review by yubiuser — CHANGES_REQUESTED** [review]

**Comment by jpgpi250** [comment]
another one that might be useful when the need arises to compare domain entries from the gravity table (gravity.db) with domain entries from the queries view (pihole-FTL.db)

```
SELECT trim(domain, '|^')
```

example (not working on standard pihole deployments)
```
# domains that used to be allowed, now blocked
exec 3>&1 1>"${importfile}"
	pihole-FTL sqlite3 -separator ',' "${FTLdb}" << EOSQL
	ATTACH '${olddb}' AS olddb;
	SELECT timestamp, domain, status from queries WHERE 
		timestamp < '${timestamp}'
		AND status IN (2, 3, 12, 13, 14)
		AND domain IN (SELECT trim(domain, '|^') FROM olddb.new) GROUP BY domain;
EOSQL
exec 1>&3 3>&-
sudo sqlite3 "${olddb}" "DELETE FROM blocked;"
sudo pihole-FTL sqlite3 -separator ',' "${olddb}" ".mode csv" ".import '${importfile}' blocked"
sudo rm "${importfile}"
```
sqlite3 select result **with** trim (sample) - **may match** an entry in the queries view:
```
zero.dns0.eu
zetland.rm-ni.uk
zougloub.eu
zrh1-ns01.monzoon.net
zxcvb.pp.ua
```
sqlite3 select result **without** trim (sample) - **will never match** anything in the queries view:
```
||zero.dns0.eu^
||zetland.rm-ni.uk^
||zougloub.eu^
||zrh1-ns01.monzoon.net^
||zxcvb.pp.ua^
```

**Comment by dschaper** [comment]
> pihole -q looks OK

`pihole -q` has not been updated to know how to parse ABP style. It likely will not have that capability when the feature is initially released. It will match exact but not subdomains. 

```
 pihole -q 1qaz.de
 Match found in http://localhost.localdomain/DOHservers/DOHadb.txt:
   ||1qaz.de^
```

That's fine and will show the exact match.

`pihole -q sub.1quaz.de` will not show that it is part of `||1qaz.de^` until we find a way for shell to be able to walk a domain while being efficient enough.

**Comment by dschaper** [comment]
> sqlite3 select result **without** trim (sample) - **will never match** anything in the queries view:

There needs to be an understanding that ABP style entries are 'special' and not a direct match for comparison.

**Comment by jpgpi250** [comment]
@dschaper @rdwebdesign
> That's fine and will show the exact match.

pihole -q sub.1quaz.de will not show that it is part of ||1qaz.de^ until we find a way for shell to be able to walk a domain while being efficient enough.

copied over [query.sh](https://raw.githubusercontent.com/pi-hole/pi-hole/abp_query/advanced/Scripts/query.sh) from branch abp_query
FYI 2.341.067 domains in gravity.

result (success):
```
pi@raspberrypi:~/tmp $ time pihole -q zzxedr.xyz
 Match found in https://v.firebog.net/hosts/AdguardDNS.txt:
   zzxedr.xyz
 Match found in https://small.oisd.nl:
   ||zzxedr.xyz^

real    0m5.325s
user    0m4.923s
sys     0m0.394s
pi@raspberrypi:~/tmp $ time pihole -q sub.zzxedr.xyz
 Match found in https://small.oisd.nl:
   ||zzxedr.xyz^

real    0m7.638s
user    0m7.324s
sys     0m0.337s
```

will keep testing...

**Comment by jpgpi250** [comment]
1. the FTL log shows a lot of entries like (partial only, there are more entries):
```
[2023-03-03 09:35:29.251 9201/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:35:29.257 9202/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:35:29.261 9203/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:35:29.771 9205/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:35:29.774 9204/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:35:29.775 9206/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:38:25.528 9431/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:38:25.531 9432/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:38:25.535 9433/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:39:35.562 9480/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:39:35.568 9481/F21551] Using ABP-style entries in gravity database
[2023-03-03 09:39:35.571 9482/F21551] Using ABP-style entries in gravity database
```
is this intended? I know I asked for a warning, but this looks like overkill. The warning, when pihole-FTL starts is sufficient.

2. what does '**9482/F21551**' and similar entries (see below, 21551/T21567) indicate in this message?
```
[2023-03-03 09:40:00.086 21551/T21567] Notice: Database size is 3.92 MB, deleted 168 rows
```

**Comment by dschaper** [comment]
> 2\. '**9482/F21551**' and similar entries (see below, 21551/T21567

I'm pretty sure those are parent PIDs and `F`ork | `T`hread PIDs.

**Comment by dschaper** [comment]
Are there any outstanding issues that need to be addressed before we can merge this in? 

Are the timing calls and the header includes meant to be removed?

**Comment by DL6ER** [comment]
> the FTL log shows a lot of entries

They should only be shown when gravity is reloaded. However, you post made me aware of that gravity *is* reloaded when forks are created as SQLite3 doesn't have proper multi-processing support when it comes to forks (only threads are fine). I'll see how this can be improved, setting back to draft until then.

@dschaper We can remove them, I'll do this as well.

**Comment by DL6ER** [comment]
I removed the leftover traces of the benchmarking we did in between and decided to also remove the ABP-style blocking hint. Ensuring it is logged for every gravity rebuild but not when, e.g., FTL forks a dedicated TCP handler seems unnecessarily complicated given that users (a) should know when they add lists with such domains, (b) `pihole -g` would tell them pretty obvious, (c) it can be checked at any time using `SELECT value FROM info WHERE property = 'abp_domains';`, and, finally, (d) it doesn't actually matter.

This PR should now finally be ready for review + merge.

**Comment by jpgpi250** [comment]
latest version installed
FTL version is new/adb_style_blocking vDev-da118e8 (Latest: v5.21)
will report back...

**Comment by jpgpi250** [comment]
- FTL behaves as expected (no problems)
- I've merged the development version of query.sh (fix for [#5199](https://github.com/pi-hole/pi-hole/issues/5199)) with the latest version of query.sh from branch [abp_query](https://github.com/pi-hole/pi-hole/tree/abp_query) (fixes match for abp_style sub domains e.g. pihole -q **sub**.zzxedr.xyz with gravity entry **||zzxedr.xyz^**). Everything appears to be working as expected.
- there is a difference in gravity.sh beween branches abp and abp_query (first line is from abp branch):
```
  valid_domain_pattern="([a-z0-9]([a-z0-9_-]{0,61}[a-z0-9]){0,1}\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]\.?"
  valid_domain_pattern="([a-z0-9]([a-z0-9_-]{0,61}[a-z0-9]){0,1}\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]"
```
see issue [4701](https://github.com/pi-hole/pi-hole/issues/4701), PR [4704](https://github.com/pi-hole/pi-hole/pull/4704/files), related?

**Comment by rdwebdesign** [comment]
> * there is a difference in gravity.sh beween branches abp and abp_query (first line is from abp branch):
> 
> ```
>   valid_domain_pattern="([a-z0-9]([a-z0-9_-]{0,61}[a-z0-9]){0,1}\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]\.?"
>   valid_domain_pattern="([a-z0-9]([a-z0-9_-]{0,61}[a-z0-9]){0,1}\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]"
> ```

This difference is expected since my code on `abp_query` is based on a previous version of `abp` branch. I created and it was initially intended as a test. I created a test branch in Feb 27. This _trailing dot_ was [introduced on Mar 4](https://github.com/pi-hole/pi-hole/commit/0b5da9f0dabccaf902aada60527c3a4b5855b925) and I didn't had the time to update my branch.

I appreciate your enthusiasm, but as I said before `abp_query` branch is still a work in progress (and there is no PR yet). 
That branch will be updated after `abp` PR is merged. 
There is no reason to rush things.

**Review by yubiuser — APPROVED** [review]

**Pull request merged**